params/json.libevm.go

```diff
@@ -39,36 +39,54 @@ type chainConfigWithExportedExtra struct {
 	Extra                      *pseudo.Type `json:"extra"` // `c.extra` is otherwise unexported
 }
 
-// UnmarshalJSON implements the [json.Unmarshaler] interface.
-func (c *ChainConfig) UnmarshalJSON(data []byte) error {
-	switch reg := registeredExtras; {
-	case reg.Registered() && !reg.Get().reuseJSONRoot:
-		return c.unmarshalJSONWithExtra(data)
-
-	case reg.Registered() && reg.Get().reuseJSONRoot: // although the latter is redundant, it's clearer
-		c.extra = reg.Get().newChainConfig()
-		if err := json.Unmarshal(data, c.extra); err != nil {
-			c.extra = nil
-			return err
-		}
-		fallthrough // Important! We've only unmarshalled the extra field.
-	default: // reg == nil
+// UnmarshalJSON implements the [json.Unmarshaler] interface and JSON decodes
+// `data` according to the following:
+//   - extra is not registered:
+//     `data` is decoded into `c` and the extra is ignored.
+//   - extra is registered and the registered reuseJSONRoot field is false:
+//     `data` is decoded into `c` and the "extra" JSON field in `data` is decoded into `c.extra`.
+//   - extra is registered and the registered reuseJSONRoot field is true:
+//     `data` is decoded into `c` and `data` is decoded into `c.extra`.
+func (c *ChainConfig) UnmarshalJSON(data []byte) (err error) {
+	if !registeredExtras.Registered() {
+		// assume there is no extra
 		return json.Unmarshal(data, (*chainConfigWithoutMethods)(c))
 	}
+	extraConstructors := registeredExtras.Get()
+	c.extra = extraConstructors.newChainConfig()
+	reuseJSONRoot := extraConstructors.reuseJSONRoot
+	return UnmarshalChainConfigJSON(data, c, c.extra, reuseJSONRoot)
 }
 
-// unmarshalJSONWithExtra unmarshals JSON under the assumption that the
-// registered [Extras] payload is in the JSON "extra" key. All other
-// unmarshalling is performed as if no [Extras] were registered.
-func (c *ChainConfig) unmarshalJSONWithExtra(data []byte) error {
-	cc := &chainConfigWithExportedExtra{
-		chainConfigWithoutMethods: (*chainConfigWithoutMethods)(c),
-		Extra:                     registeredExtras.Get().newChainConfig(),
-	}
-	if err := json.Unmarshal(data, cc); err != nil {
-		return err
+// UnmarshalChainConfigJSON JSON decodes `data` according to the following.
+//   - `extra` is not nil and `reuseJSONRoot` is false:
+//     `data` is decoded into `config` and the "extra" JSON field in `data` is decoded into `extra`.
+//   - `extra` is not nil and `reuseJSONRoot` is true:
+//     `data` is decoded into `config` and `data` is decoded into `extra`.
+//   - `extra` is nil:
+//     `data` is decoded into `config` and the extra is ignored.
+func UnmarshalChainConfigJSON[T any](data []byte, config *ChainConfig, extra *T, reuseJSONRoot bool) (err error) {
+	err = json.Unmarshal(data, (*chainConfigWithoutMethods)(config))
+	switch {
+	case err != nil:
+		return fmt.Errorf("decoding root chain config: %s", err)
+	case extra == nil: // ignore the extra, whether at the root JSON depth, or at the "extra" JSON field.
```

Comment:
do we want to support this case where the user pass in a `nil` extra? 🤔 Or not really?

Reply:
I changed to return an error `extra pointer argument is nil` because:

- I don't think there is a point using this function at all if the extra argument is a (typed) nil
- Previously, it was inconsistent behavior: it would error when re-using the JSON root (from `json.Unmarshal(data, extra)`) but would silently discard the "extra" JSON field without error if not re-using the JSON root.
- Discarding silently the extra field sounds like a dangerous path, better to return an error.

Reply:
No, we don't want to support nil extras because that is never a valid use case. In the extras-registered scenario it will never be nil (by design), and in the non-registered scenario the caller should just use the regular JSON decoding mechanism of either `json.Decode([]byte, *ChainConfig)` or `ChainConfig.UnmarshalJSON()`.

Reply:
Yes I think returning an error in case the extra is nil makes sense in that case then.